File: Problems 1900-1999/1980. Find Unique Binary String/solution.cpp
class Solution {
public:
    string findDifferentBinaryString(vector<string>& nums) {
        int n = nums.size();

        // differ from each string by at least one bit
        string res;
        for (int i = 0; i < n; i++) {
            res += nums[i][i] == '0' ? '1' : '0';
        }
        return res;

        // int n = nums.size();
        // unordered_set<string> binaries(nums.begin(), nums.end());

        // string curr;
        // return findBinaryString(0, n, binaries, curr);
    }

    string findBinaryString(int i, int& n, unordered_set<string>& nums, string& curr) {
        if (i == n) return nums.find(curr) == nums.end() ? curr : "";
        // 0
        curr.push_back('0');
        string res = findBinaryString(i+1, n, nums, curr);
        if (res.length() == n) return res;
        curr.pop_back();
        // 1
        curr.push_back('1');
        res = findBinaryString(i+1, n, nums, curr);
        if (res.length() == n) return res;
        curr.pop_back();
        return "";
    }
};

File: Problems 1900-1999/1980. Find Unique Binary String/solution.java
class Solution {
    public String findDifferentBinaryString(String[] nums) {
        // Cantor's diagonalization
        // Since we want to create a unique binary string of length n,
        // and there are n strings in nums, we can just iterate over
        // each string in nums and append a different bit per string
        // to construct the binary string of length n
        // Since we know the string differs from each string in nums by
        // at least one bit, we know it is unique
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < nums.length; i++) {
            sb.append(nums[i].charAt(i) == '0' ? '1' : '0');
        }

        return sb.toString();
    }
}